core/src/main/java/com/linecorp/armeria/common/util/MoreSystem.java

```diff
@@ -0,0 +1,224 @@
+            Process process = null;
+            String hostname = null;
+            try {
+                process = Runtime.getRuntime().exec("hostname");
+                final BufferedReader in = new BufferedReader(new InputStreamReader(process.getInputStream()));
+                final String line = in.readLine();
+                if (line == null) {
+                    logger.debug("The 'hostname' command returned nothing; " +
+                                 "using InetAddress.getLocalHost() instead");
+                } else {
+                    hostname = normalizeHostname(line);
+                    if (hostname == null) {
+                        logger.debug("The 'hostname' command returned a non-hostname ({}); " +
+                                     "using InetAddress.getLocalHost() instead", line);
+                    } else {
+                        logger.info("Hostname: {} (from 'hostname' command)", hostname);
+                    }
+                }
+            } catch (Exception e) {
+                logger.debug("Failed to get the hostname using the 'hostname' command; " +
+                             "using InetAddress.getLocalHost() instead", e);
+            } finally {
+                if (process != null) {
+                    process.destroy();
+                }
+            }
+
+            if (hostname == null) {
+                try {
+                    final String jdkHostname = InetAddress.getLocalHost().getHostName();
+                    hostname = normalizeHostname(jdkHostname);
+                    if (hostname == null) {
+                        logger.warn("InetAddress.getLocalHost() returned an invalid hostname ({}); " +
+                                    "using 'localhost' instead", jdkHostname);
+                    } else {
+                        logger.info("Hostname: {} (from InetAddress.getLocalHost())", hostname);
+                    }
+                } catch (Exception e) {
+                    logger.warn("Failed to get the hostname using InetAddress.getLocalHost(); " +
+                                "using 'localhost' instead", e);
+                }
+            }
+
+            HOSTNAME = firstNonNull(hostname, "localhost");
+        }
+
+        @Nullable
+        private static String normalizeHostname(String line) {
+            final String hostname = IDN.toASCII(line.trim(), IDN.ALLOW_UNASSIGNED);
+            if (!HOSTNAME_PATTERN.matcher(hostname).matches()) {
+                return null;
+            }
+            return Ascii.toLowerCase(hostname);
+        }
+    }
+
+    private static final class Pid {
+
+        static final int PID;
+
+        static {
+            int pid = -1;
+            // Try ProcessHandle.pid() if Java 9+.
+            if (PlatformDependent.javaVersion() >= 9) {
+                try {
+                    final Class<?> handleClass = Class.forName("java.lang.ProcessHandle", true,
```

Comment:
What do you think about using class loading instead of reflection? I guess it might complicate the build though

Reply:
Yeah, `ProcessHandle` is not available in Java 8 so it will complicate the build, especially that we pass `--release 8` to the compiler universally. Let me keep this because it simplifies the build. :-)